alibi_detect/utils/pytorch/kernels.py

```diff
@@ -1,8 +1,44 @@
+from abc import abstractmethod
 import numpy as np
 import torch
 from torch import nn
 from . import distance
-from typing import Optional, Union, Callable
+from typing import Optional, Union, Callable, List
+from copy import deepcopy
+
+
+def infer_kernel_parameter(
+    kernel: 'BaseKernel',
+    x: torch.Tensor,
+    y: torch.Tensor,
+    dist: torch.Tensor,
+    infer_parameter: bool = True
+) -> None:
+    """
+    Infer the kernel parameter from the data.
+
+    Parameters
+    ----------
+    kernel
+        The kernel function.
+    x
+        Tensor of instances with dimension [Nx, features].
+    y
+        Tensor of instances with dimension [Ny, features].
+    dist
+        Tensor with dimensions [Nx, Ny], containing the pairwise distances between `x` and `y`.
+    infer_parameter
+        Whether to infer the kernel parameter.
+    """
+    if kernel.trainable and infer_parameter:
+        raise ValueError("Gradients cannot be computed w.r.t. an inferred sigma value")
+    for parameter in kernel.parameter_dict.values():
```

Comment:
Is there a specific reason why we can't adhere to the usual way of initialising torch parameters, with custom init functions?

Reply:
Here I just followed the previous error message, which doesn't allow both `trainable` and `infer_parameter`. Should be easily changed if allowing both is the desired behaviour.